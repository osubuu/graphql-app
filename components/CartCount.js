import PropTypes from 'prop-types';
import styled from 'styled-components';
import { TransitionGroup, CSSTransition } from 'react-transition-group';

const Dot = styled.div`
  background: ${props => props.theme.red};
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;

  /* keep character size of each number constant
  (e.g. skinny '1' will keep same width as larger '2) */
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  /* Initial state of the entered Dot (flip on its back) */
  .count-enter {
    /* 0.5turn = 180 degrees */
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    position: absolute;
    top: 0;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

const CartCount = ({ count }) => (
  <AnimationStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 400, exit: 400 }}
      >
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyles>
);

CartCount.propTypes = {
  count: PropTypes.number.isRequired,
};

export default CartCount;
