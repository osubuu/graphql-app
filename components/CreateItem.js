import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';

import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import { PAGINATION_QUERY } from './Pagination';

// recall: these mutations and queries were defined in schema.grapql yoga file in the back-end
export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  }

  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === 'number' && value ? parseFloat(value) : value;
    this.setState({ [name]: val });
  }

  uploadFile = async e => {
    const { files } = e.target;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'graphql-app');

    const res = await fetch('https://api.cloudinary.com/v1_1/df7nqbjav/image/upload/', {
      method: 'POST',
      body: data,
    });

    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    });
  }

  render() {
    return (
      <Mutation
        mutation={CREATE_ITEM_MUTATION}
        refetchQueries={[
          { query: PAGINATION_QUERY },
        ]}
        variables={this.state}
      >
        {(createItem, { loading, error }) => (
          <Form
            data-test="form"
            onSubmit={async e => {
              e.preventDefault();
              const res = await createItem(); // sends the mutation request
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              });
            }}
          >
            <ErrorMessage error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  required
                  placeholder="Upload an image"
                  onChange={this.uploadFile}
                />
                {this.state.image && (
                <img
                  width="200"
                  src={this.state.image}
                  alt="Upload Preview"
                />
                )}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  required
                  placeholder="Title"
                  value={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  required
                  placeholder="Price"
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <input
                  id="description"
                  name="description"
                  required
                  placeholder="Enter a description"
                  value={this.state.description}
                  onChange={this.handleChange}
                />
              </label>
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
