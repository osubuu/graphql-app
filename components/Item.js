import PropTypes from 'prop-types';
import Link from 'next/link';

import Title from './styles/Title';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';

import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';

const Item = props => {
  const { item, me } = props;
  const canEditAndDelete = me && me.id === item.user.id;
  return (
    <ItemStyles>
      <Link href={{
        pathname: '/item',
        query: { id: item.id },
      }}
      >
        <a>
          {item.image && <img src={item.image} alt={item.title} />}
        </a>
      </Link>
      <Title>
        <Link href={{
          pathname: '/item',
          query: { id: item.id },
        }}
        >
          <a>{item.title}</a>
        </Link>
      </Title>
      <PriceTag>{formatMoney(item.price)}</PriceTag>
      <p>{item.description}</p>

      <div className="buttonList">
        {canEditAndDelete && (
        <Link href={{
          pathname: 'update',
          query: { id: item.id },
        }}
        >
          <a>Edit</a>
        </Link>
        )}
        <AddToCart id={item.id} />
        {canEditAndDelete && <DeleteItem id={item.id}>Delete This Item</DeleteItem>}
      </div>
    </ItemStyles>
  );
};

Item.propTypes = {
  item: PropTypes.oneOfType([PropTypes.object]).isRequired,
};

export default Item;
