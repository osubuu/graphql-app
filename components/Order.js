import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import { format } from 'date-fns';
import Head from 'next/head';
import gql from 'graphql-tag';
import formatMoney from '../lib/formatMoney';
import ErrorMessage from './ErrorMessage';
import OrderStyles from './styles/OrderStyles';

export const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      id
      charge
      total
      createdAt
      user {
        id
      }
      items {
        id
        title
        description
        price
        image
        quantity
      }
    }
  }
`;

const Order = props => (
  <Query query={SINGLE_ORDER_QUERY} variables={{ id: props.id }}>
    {({ data, error, loading }) => {
      if (error) return <ErrorMessage error={error} />;
      if (loading) return <p>Loading...</p>;
      const { order } = data;
      return (
        <OrderStyles data-test="order">
          <Head>
            <title>Sick Fits - Order {order.id}</title>
          </Head>
          <p>
            <span>Order ID:</span>
            <span>{order.id}</span>
          </p>
          <p>
            <span>Charge</span>
            <span>{order.charge}</span>
          </p>
          <p>
            <span>Date</span>
            <span>{format(order.createdAt, 'MMMM d, YYYY h:mm a')}</span>
          </p>
          <p>
            <span>Order Total</span>
            <span>{formatMoney(order.total)}</span>
          </p>
          <p>
            <span>Item Count</span>
            <span>{order.items.length}</span>
          </p>
          <div className="items">
            {order.items.map(item => (
              // TODO: make it its own component
              <div className="order-item" key={item.id}>
                <img src={item.image} alt={item.title} />
                <div className="item-details">
                  <h2>{item.title}</h2>
                  <p>Qty: {item.quantity}</p>
                  <p>Each: {formatMoney(item.price)}</p>
                  <p>Subtotal: {formatMoney(item.price * item.quantity)}</p>
                  <p>{item.description}</p>
                </div>
              </div>
            ))}
          </div>
        </OrderStyles>
      );
    }}
  </Query>
);

Order.propTypes = {
  id: PropTypes.string.isRequired,
};

export default Order;
