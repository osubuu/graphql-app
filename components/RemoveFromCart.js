import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { CURRENT_USER_QUERY } from './User';

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${props => props.theme.red};
    cursor: pointer;
  }
`;

class RemoveFromCart extends Component {
  // this gets called as soon as we get a response back from the server
  // after a mutation has been performed
  update = (cache, payload) => {
    // 1. First read the cache
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });
    // 2. Remove that item from the cart
    const cartItemId = payload.data.removeFromCart.id;
    data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);
    // 3. Write it back to the cache
    cache.writeQuery({ query: CURRENT_USER_QUERY, data });
  }

  render() {
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
        // run a function preemptively ASSUMING the mutation would be a success
        // this is for optimal UX
        optimisticResponse={{
          __typename: 'Mutation',
          removeFromCart: {
            __typename: 'CartItem',
            id: this.props.id,
          },
        }}
      >
        {(removeFromCart, { loading }) => (
          <BigButton
            title="Delete Item"
            disabled={loading}
            onClick={() => {
              removeFromCart().catch(err => alert(err.message));
            }}
          >
            &times;
          </BigButton>
        )}
      </Mutation>
    );
  }
}

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired,
};

export default RemoveFromCart;
