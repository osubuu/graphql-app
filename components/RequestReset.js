import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';

import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import { Fieldset } from './Signin';

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

class RequestReset extends Component {
  state = {
    email: '',
  }

  saveToState = e => {
    this.setState({ [e.target.name]: e.target.value });
  }

  render() {
    return (
      <Mutation
        mutation={REQUEST_RESET_MUTATION}
        variables={this.state}
      >
        {(reset, { error, loading, called }) => (
          <Form
            data-test="form"
            method="post"
            onSubmit={async e => {
              e.preventDefault();
              await reset();
              this.setState({ email: '' });
            }}
          >
            <Fieldset disabled={loading} aria-busy={loading}>
              <div>
                <h2>Request a password reset</h2>
                <ErrorMessage error={error} />
                {!error && !loading && called && <p>Success! Check your email for a reset link!</p>}
                <label htmlFor="email">
                  Email
                  <input
                    type="email"
                    name="email"
                    placeholder="email"
                    value={this.state.email}
                    onChange={this.saveToState}
                  />
                </label>
                <button type="submit">Request reset</button>
              </div>
            </Fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default RequestReset;
