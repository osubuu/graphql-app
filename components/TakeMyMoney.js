import React, { Component } from 'react';
import StripeCheckout from 'react-stripe-checkout';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import NProgress from 'nprogress';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import calcTotalPrice from '../lib/calcTotalPrice';
import ErrorMessage from './ErrorMessage';
import User, { CURRENT_USER_QUERY } from './User';

const CREATE_ORDER_MUTATION = gql`
  mutation createOrder($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

const totalItems = cart => cart.reduce((total, cartItem) => total + cartItem.quantity, 0);

class TakeMyMoney extends Component {
  // get token to then pass it to backend to actually charge the card
  // (i.e token here is sort of like a meal ticket)
  onToken = async (res, createOrder) => {
    NProgress.start();
    // manually call the mutation once we have the stripe token
    const order = await createOrder({
      variables: {
        token: res.id,
      },
    }).catch(err => { alert(err.message) });

    Router.push({
      pathname: '/order',
      query: { id: order.data.createOrder.id },
    });
  }

  render() {
    return (
      <User>
        {({ data: { me } }) => (
          <Mutation mutation={CREATE_ORDER_MUTATION} refetchQueries={[{
            query: CURRENT_USER_QUERY,
          }]}>
            {createOrder => (
              <StripeCheckout
                name="Sick Fits"
                email={me.email}
                token={res => this.onToken(res, createOrder)}
                image={me.cart.length && me.cart[0].item && me.cart[0].item.image}
                amount={calcTotalPrice(me.cart)}
                currency="USD"
                stripeKey="pk_test_6yUD5q8liiWCZPF78ZQZe3Ie00ixD2vxhr"
                description={`Order of ${totalItems(me.cart)} items`}>
                <p>{this.props.children}</p>
              </StripeCheckout>
            )}
          </Mutation>
        )}
      </User>
    );
  }
}

export default TakeMyMoney;
