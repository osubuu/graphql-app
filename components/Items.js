import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import { adopt } from 'react-adopt';

import Pagination from './Pagination';
import User from './User';
import Item from './Item';
import { perPage } from '../config';

// best practice to put queries in CAPS
export const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
    items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
      user {
        id
      }
    }
  }
`;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`;


const Items = props => {
  const Composed = adopt({
    user: ({ render }) => <User>{render}</User>,
    items: ({ render }) => (
      <Query
        query={ALL_ITEMS_QUERY}
        variables={{
          skip: props.page * perPage - perPage,
        }}
      >{render}
      </Query>
    ),
  });

  return (
    <Composed>
      {({ user, items }) => {
        if (items.loading) return <p>Loading...</p>;
        if (items.error) return <p>Error: {items.error.message}</p>;
        return (
          <Center>
            <Pagination page={props.page} />
            <ItemsList>
              {items.data.items.map(item => <Item item={item} key={item.id} me={user.data.me} />)}
            </ItemsList>
            <Pagination page={props.page} />
          </Center>
        );
      }}
    </Composed>
  );
};

export default Items;
