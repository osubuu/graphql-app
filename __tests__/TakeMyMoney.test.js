import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import NProgress from 'nprogress';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';

import TakeMyMoney from '../components/TakeMyMoney';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';

Router.router = { push() {} };

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem()],
        },
      },
    },
  },
];

describe('<TakeMyMoney/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney />
      </MockedProvider>,
    );
    await wait();
    wrapper.update();
    const checkoutButton = wrapper.find('ReactStripeCheckout');
    expect(toJSON(checkoutButton)).toMatchSnapshot();
  });

  it('creates an order on token', async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: { id: 'xyz789' },
      },
    });
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney />
      </MockedProvider>,
    );
    const component = wrapper.find('TakeMyMoney').instance();
    // manually call the onToken method
    component.onToken({ id: 'abc123' }, createOrderMock);
    expect(createOrderMock).toHaveBeenCalledWith({ variables: { token: 'abc123' } });
  });

  it('turns the progress bar on', async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: { id: 'xyz789' },
      },
    });
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney />
      </MockedProvider>,
    );
    await wait();
    wrapper.update();
    const component = wrapper.find('TakeMyMoney').instance();
    NProgress.start = jest.fn();
    component.onToken({ id: 'abc123' }, createOrderMock);
    expect(NProgress.start).toHaveBeenCalled();
  });

  it('routes to the order page when completed', async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: { id: 'xyz789' },
      },
    });
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney />
      </MockedProvider>,
    );
    await wait();
    wrapper.update();
    const component = wrapper.find('TakeMyMoney').instance();
    component.onToken({ id: 'abc123' }, createOrderMock);
    Router.router.push = jest.fn();
    await wait();
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: '/order',
      query: { id: 'xyz789' },
    });
  });
});
