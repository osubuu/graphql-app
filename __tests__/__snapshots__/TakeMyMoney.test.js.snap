// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TakeMyMoney/> renders and matches snapshot 1`] = `
<ReactStripeCheckout
  ComponentClass="span"
  amount={15000}
  className="StripeCheckout"
  currency="USD"
  description="Order of 3 items"
  email="Delmer.Smith@yahoo.com"
  image="dog-small.jpg"
  label="Pay With Card"
  locale="auto"
  name="Hoodify"
  reconfigureOnUpdate={false}
  stripeKey="pk_test_6yUD5q8liiWCZPF78ZQZe3Ie00ixD2vxhr"
  token={[Function]}
  triggerEvent="onClick"
>
  <span
    onClick={[Function]}
  >
    <p />
  </span>
</ReactStripeCheckout>
`;
